src/direct.js

```diff
@@ -0,0 +1,468 @@
+import { Queue } from './common/utils';
+import * as utils from './common/utils';
+import * as constants from './common/constants';
+import WavefrontSDKMetricsRegistry from './common/metrics/registry';
+
+require('isomorphic-fetch');
+const pako = require('pako');
+
+/**
+ * Wavefront direct ingestion client.
+ * Sends data directly to Wavefront cluster via the direct ingestion API.
+ */
+export default class WavefrontDirectClient {
+  /**
+   * Construct Direct Client.
+   * @param server - Server address, Example: https://INSTANCE.wavefront.com
+   * @param token - Token with Direct Data Ingestion permission granted
+   * @param maxQueueSize - Size of internal data buffer for each data type
+   * @param batchSize - Amount of data sent by one api call
+   * @param flushIntervalSeconds - Interval flush interval
+   * @param enableInternalMetrics
+   */
+  constructor({
+    server,
+    token,
+    maxQueueSize = 50000,
+    batchSize = 10000,
+    flushIntervalSeconds = 5,
+    enableInternalMetrics = true
+  }) {
+    this.server = server;
+    this._batchSize = batchSize;
+    this._flushIntervalSeconds = flushIntervalSeconds;
+    this._defaultSource = null;
+    this._metricsBuffer = new Queue(maxQueueSize);
+    this._histogramsBuffer = new Queue(maxQueueSize);
+    this._tracingSpansBuffer = new Queue(maxQueueSize);
+    this._spanLogsBuffer = new Queue(maxQueueSize);
+    this._headers = {
+      'Content-Type': 'application/octet-stream',
+      'Content-Encoding': 'gzip',
+      Authorization: 'Bearer ' + token
+    };
+    this._timer = null;
+    this._scheduleTimer();
+
+    if (enableInternalMetrics) {
+      this._sdkMetricsRegistry = new WavefrontSDKMetricsRegistry({
+        wfMetricSender: this,
+        prefix: `${constants.SDK_METRIC_PREFIX}.core.sender.direct`
+      });
+    } else {
+      this._sdkMetricsRegistry = new WavefrontSDKMetricsRegistry({
+        wfMetricSender: null
+      });
+    }
+
+    this._sdkMetricsRegistry.newGauge(
+      'points.queue.size',
+      this._metricsBuffer.size
+    );
+    this._sdkMetricsRegistry.newGauge(
+      'points.queue.remaining_capacity',
+      this._metricsBuffer.remainCapacity
+    );
+    this._pointsValid = this._sdkMetricsRegistry.newCounter('points.valid');
+    this._pointsInvalid = this._sdkMetricsRegistry.newCounter('points.invalid');
+    this._pointsDropped = this._sdkMetricsRegistry.newCounter('points.dropped');
+    this._pointsReportErrors = this._sdkMetricsRegistry.newCounter(
+      'points.report.errors'
+    );
+
+    this._sdkMetricsRegistry.newGauge(
+      'histograms.queue.size',
+      this._histogramsBuffer.size
+    );
+    this._sdkMetricsRegistry.newGauge(
+      'histograms.queue.remaining_capacity',
+      this._histogramsBuffer.remainCapacity
+    );
+    this._histogramsValid = this._sdkMetricsRegistry.newCounter(
```

Comment:
can we make this data-driven?  This code is very repetitive.
```
const histogramMapping = { _histogramsInvalid : "histograms.invalid", ... etc }
Object.entries(histogramMapping).forEach(([key, value]) => { this[key] = this._sdkMetricsRegistry.newCounter(value) })
```

Similar constructs can be applied across the class.  Leave it to you to decide what level of complexity you're comfortable with.

Reply:
thanks for the suggestion! changed the lines here